{"cells":[{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["import subprocess\n","import os\n","input_fasta = \"/Users/xaviersanterre/test/BP.fasta\" "]},{"cell_type":"code","execution_count":6,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["\n","\n","     ______  ____   ____  ____   ____  ______  __ __\n","    |      ||    \\ |    ||    \\ |    ||      ||  |  |\n","    |      ||  D  ) |  | |  _  | |  | |      ||  |  |\n","    |_|  |_||    /  |  | |  |  | |  | |_|  |_||  ~  |\n","      |  |  |    \\  |  | |  |  | |  |   |  |  |___, |\n","      |  |  |  .  \\ |  | |  |  | |  |   |  |  |     |\n","      |__|  |__|\\_||____||__|__||____|  |__|  |____/\n","\n","    Trinity-v2.15.1\n","\n","\n","\n","Single read files: $VAR1 = [\n","          '/Users/xaviersanterre/test/BP.fasta'\n","        ];\n","Trinity version: Trinity-v2.15.1\n","-currently using the latest production release of Trinity.\n","\n","Error, cannot find samtools. Please be sure samtools is installed and included in your PATH setting.\n"]},{"ename":"CalledProcessError","evalue":"Command '['/Users/xaviersanterre/Test/trinityrnaseq-v2.15.1/Trinity', '--seqType', 'fq', '--single', '/Users/xaviersanterre/test/BP.fasta', '--CPU', '2', '--max_memory', '1G', '--output', '/Users/xaviersanterre/test']' returned non-zero exit status 1.","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)","Cell \u001b[0;32mIn[6], line 21\u001b[0m\n\u001b[1;32m     18\u001b[0m input_fasta \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/Users/xaviersanterre/test/BP.fasta\u001b[39m\u001b[38;5;124m\"\u001b[39m \n\u001b[1;32m     19\u001b[0m output_directory \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/Users/xaviersanterre/test\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m---> 21\u001b[0m \u001b[43mrun_trinity\u001b[49m\u001b[43m(\u001b[49m\u001b[43minput_fasta\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moutput_directory\u001b[49m\u001b[43m)\u001b[49m\n","Cell \u001b[0;32mIn[6], line 15\u001b[0m, in \u001b[0;36mrun_trinity\u001b[0;34m(input_fasta, output_dir, max_memory, cpu_count)\u001b[0m\n\u001b[1;32m      4\u001b[0m trinity_path \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/Users/xaviersanterre/Test/trinityrnaseq-v2.15.1/Trinity\u001b[39m\u001b[38;5;124m'\u001b[39m  \u001b[38;5;66;03m# Replace with the actual full path\u001b[39;00m\n\u001b[1;32m      6\u001b[0m trinity_command \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m      7\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/Users/xaviersanterre/Test/trinityrnaseq-v2.15.1/Trinity\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m      8\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m--seqType\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfq\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     12\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m--output\u001b[39m\u001b[38;5;124m\"\u001b[39m, output_dir\n\u001b[1;32m     13\u001b[0m ]\n\u001b[0;32m---> 15\u001b[0m \u001b[43msubprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtrinity_command\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcheck\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n","File \u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py:524\u001b[0m, in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    522\u001b[0m     retcode \u001b[38;5;241m=\u001b[39m process\u001b[38;5;241m.\u001b[39mpoll()\n\u001b[1;32m    523\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m check \u001b[38;5;129;01mand\u001b[39;00m retcode:\n\u001b[0;32m--> 524\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m CalledProcessError(retcode, process\u001b[38;5;241m.\u001b[39margs,\n\u001b[1;32m    525\u001b[0m                                  output\u001b[38;5;241m=\u001b[39mstdout, stderr\u001b[38;5;241m=\u001b[39mstderr)\n\u001b[1;32m    526\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m CompletedProcess(process\u001b[38;5;241m.\u001b[39margs, retcode, stdout, stderr)\n","\u001b[0;31mCalledProcessError\u001b[0m: Command '['/Users/xaviersanterre/Test/trinityrnaseq-v2.15.1/Trinity', '--seqType', 'fq', '--single', '/Users/xaviersanterre/test/BP.fasta', '--CPU', '2', '--max_memory', '1G', '--output', '/Users/xaviersanterre/test']' returned non-zero exit status 1."]}],"source":["#Add coding lines for installation of samtools, jellyfish, bowtie2, salmon as there are required for trinity to run. \n","\n","def run_trinity(input_fasta_path, output_dir_path, max_memory=\"1G\", cpu_count=\"2\",FASTQC): ##this mightneed to be a memory of 50 G and CPU of like 10 if we use all pooled data like Jasmin said.\n","    \n","     # Ask the person to select stranded, unstranded or reverse stranded\n","    check = True # This boolean will be switched off once the person enters a valid input\n","    while check:\n","        strandedness = input([\"Please input whether you're RNA-seq library is \\r\", \n","                            \"(y) forward/sense stranded \\r\",\n","                            \"(n) unstranded or \\r\",\n","                            \"(reverse) reverse stranded \\r\", \n","                            \"Type y, n, or reverse: \"])\n","        # Forward stranded = the read represents the RNA sequence directly (complement of the template DNA strand)\n","        # Reverse stranded = the read represents the complement of the RNA sequence\n","        # Unstranded = there is no info on whether it is forward or reverse stranded\n","\n","        strandedness = strandedness.lower() # convert person's answer to lowercase\n","\n","        # Check that the input is valid\n","        if strandedness == \"y\" or strandedness == \"n\" or strandedness == \"reverse\":\n","            check = False\n","\n","        if strandedness == \"n\":\n","             trinity_command = [\n","            \"trinity\",\n","            \"--seqType\", \"fq\",\n","            \"--single\", input_fasta_path,\n","            \"--CPU\", cpu_count,\n","            \"--max_memory\", max_memory,\n","            \"--output\", output_dir_path\n","            ]\n","\n","        else:       \n","            #write a line to ask for FASTQ file? and then seperate in R1 and R2\n","            \n","\n","            if strandedness == \"y\":\n","                #write a line to assign R1 and R2\n","                #left_reads = \n","                #right_reads\n","\n","\n","                trinity_command = [\n","                \"Trinity\",\n","                \"--seqType\", \"fq\",\n","                '--left', left_reads, \n","                '--right', right_reads, \n","                '--SS_lib_type', 'FR',  # Specify strand-specific library type,\n","                \"--CPU\", cpu_count,\n","                \"--max_memory\", max_memory,\n","                \"--output\", output_dir_path\n","                ]\n","\n","            else:\n","                #write a line to assign R1 and R2\n","                #left_reads = \n","                #right_reads\n","\n","\n","                trinity_command = [\n","                \"Trinity\",\n","                \"--seqType\", \"fq\",\n","                '--left', left_reads, \n","                '--right', right_reads, \n","                '--SS_lib_type', 'RF',  # Specify strand-specific library type,\n","                \"--CPU\", cpu_count,\n","                \"--max_memory\", max_memory,\n","                \"--output\", output_dir_path\n","                ]\n","\n","'''\n","    trinity_command = [\n","        \"trinity\",\n","        \"--seqType\", \"fq\",\n","        \"--single\", input_fasta_path,\n","        \"--CPU\", cpu_count,\n","        \"--max_memory\", max_memory,\n","        \"--output\", output_dir_path\n","    ]\n","'''\n","    \n","    trinity_result = subprocess.run(trinity_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)\n","# Deal with errors\n","    if trinity_result.returncode == 0:\n","        # Command was successful, process the output\n","        with open(output_dir_path, \"w\") as output_file:\n","            output_file.write(trinity_result.stdout)\n","    else:\n","        # There was an error\n","        print(\"Error in execution of htseq-count. Please check that the input paths are valid and that the file types are correct.\")\n","        print(trinity_result.stderr)\n","\n"]},{"cell_type":"markdown","metadata":{},"source":[]},{"cell_type":"code","execution_count":7,"metadata":{},"outputs":[{"ename":"SyntaxError","evalue":"invalid syntax (342002119.py, line 1)","output_type":"error","traceback":["\u001b[0;36m  Cell \u001b[0;32mIn[7], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    Users/xaviersanterre/Test/trinityrnaseq-v2.15.1/Trinity\u001b[0m\n\u001b[0m                                              ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"]}],"source":["#Add coding lines for installation of samtools, jellyfish, bowtie2, salmon as there are required for trinity to run. \n","\n","def run_trinity(input_fasta_path, output_dir_path, max_memory=\"1G\", cpu_count=\"2\"):\n","    trinity_path = '/Users/xaviersanterre/Test/trinityrnaseq-v2.15.1/Trinity'  # Replace with the actual full path\n","    \n","    trinity_command = [\n","        \"/Users/xaviersanterre/Test/trinityrnaseq-v2.15.1/Trinity\",\n","        \"--seqType\", \"fq\",\n","        \"--single\", input_fasta,\n","        \"--CPU\", cpu_count,\n","        \"--max_memory\", max_memory,\n","        \"--output\", output_dir_path\n","    ]\n","    \n","    subprocess.run(trinity_command, check=True)\n","\n","# Main workflow\n","input_fasta = \"/Users/xaviersanterre/test/BP.fasta\" \n","output_directory = \"/Users/xaviersanterre/test\"\n","\n","run_trinity(input_fasta, output_directory)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["#This technique would combine all reads before performing Trinity\n","\n","'''IMPORT SUBPROCESS'''\n","import subprocess\n","\n","'''COMBINE READS'''\n","#NOTE: For a single-read run, one Read 1 (R1) FASTQ file is created for each sample per flow cell lane. \n","      #For a paired-end run, one R1 and one Read 2 (R2) FASTQ file is created for each sample for each lane.\n","#If single-read run:\n","\n","\n","#If paired-end run:\n","#Write functions to concatenate forward and reverse reads in paired en case:\n","def combine_reads(reads_pattern, output_file):\n","    \"\"\"\n","    Combine reads matching a pattern into a single file. So all R1 in one file and all R2 in the other\n","    Parameters:\n","    reads_pattern (str): Glob pattern for matching read files.\n","    output_file (str): Name of the output file.\n","    \"\"\"\n","    # Using shell command to concatenate files\n","    cat_command = f\"cat {reads_pattern} > {output_file}\"\n","    subprocess.run(cat_command, shell=True, check=True)\n","    print(f\"Combined reads into {output_file}\")\n","\n"," \n","\n","'''COMBINE EVERYTHING IN MAIN FUNCTION'''\n","\n","def main():\n","    # Define your read patterns and output files\n","    forward_pattern = \"*_1.fastq\" #Here basically all the R1 will technically be denoted _1.fastq while R2 will be denoted _2fastq\n","    reverse_pattern = \"*_2.fastq\"\n","    combined_forward = \"combined_1.fastq\"\n","    combined_reverse = \"combined_2.fastq\"\n","    \n","    # Combine reads\n","    combine_reads(forward_pattern, combined_forward)\n","    combine_reads(reverse_pattern, combined_reverse)\n","    \n","    # Optionally, run FastQC on the combined files\n","    # Uncomment the next line to enable FastQC check\n","    # run_fastqc([combined_forward, combined_reverse])\n","    \n","    # Run Trinity\n","    run_trinity(combined_forward, combined_reverse)\n","\n","if __name__ == \"__main__\":\n","    main()\n","    \n","'''RUN TRINITY'''\n","\n","def run_trinity(forward_reads, reverse_reads, memory=\"50G\", cpu=\"10\"):\n","    \"\"\"\n","    Run Trinity for de novo transcriptome assembly.\n","    \n","    Parameters:\n","    forward_reads (str): File name of combined forward reads.\n","    reverse_reads (str): File name of combined reverse reads.\n","    memory (str): Maximum memory for Trinity to use.\n","    cpu (str): Number of CPU threads for Trinity.\n","    \"\"\"\n","    trinity_command = f\"Trinity --seqType fq --max_memory {memory} --left {forward_reads} --right {reverse_reads} --CPU {cpu}\"\n","    subprocess.run(trinity_command, shell=True, check=True)\n","    print(\"Trinity assembly completed\")\n","\n","'''EXECUTE'''\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["import os\n","import glob\n","\n","def concatenate_fasta_files(output_file, *input_directories):    \n","    \"\"\"\n","    Concatenate FASTA files from different conditions into a single file for proper de novo assembly with Trinity\n","\n","    Parameters:\n","    output_file (str): Path to the output FASTA file.\n","    *input_directories (str): Variable number of arguments, each specifying a directory containing FASTA files.\n","    Example: concatenate_fasta_files(output_fasta,\n","                        'path/to/condition1_fasta_files',\n","                        'path/to/condition2_fasta_files',\n","                        'path/to/condition3_fasta_files',\n","                        'path/to/condition4_fasta_files')\n","    \"\"\"\n","    # Open the output file\n","    with open(output_file, 'w') as outfile:\n","        # Process each directory of FASTA files\n","        for directory in input_directories:\n","            # Create a list of FASTA files in the current directory\n","            fasta_files = glob.glob(os.path.join(directory, '*.fasta'))\n","            print(f\"Processing {len(fasta_files)} files in {directory}\") #This line will confirm to the user the amount of files that is being processed\n","            \n","            # Iterate through each FASTA file\n","            for file in fasta_files:\n","                with open(file, 'r') as infile:\n","                    # Write each line from the input file to the output file so to concatenate all the FASTA files in all the different conditions. \n","                    outfile.write(infile.read())\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]}],"metadata":{"kernelspec":{"display_name":"Test","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.9.1"}},"nbformat":4,"nbformat_minor":2}
